### ISSUE TITLE
About how to choose weights and how to train

### ISSUE THREAD

**Issue by qmechlab** [TreeLLi/APT#5]
Hi, your paper and code are fantastic. I have two problems with the implementation.
1) Can you please tell me which weight from the provided APT weights folder should be selected for zero shot experiments. 
2) Also I used the default command line “bash scripts/APT.sh imagenet vit_b32_ep50 end 16 16 False 4 2.67 3 0 onfly 0” for training. The trained weights were then subjected to zero shot experiments, and my experiments were not very good in terms of clean accuracy. I would like to ask if the above settings need to be changed?

Thank you for all the help you can provide!

**Comment by TreeLLi** [comment]
Hi @zhangyun-code,

thanks for your interest in this work. To reproduce the zero-shot evaluation results reported in the paper, you should use the weights named `zs_imagenet_vitb32_eps4_nctx1_100shots_UC` which I just uploaded. 

Regarding your evaluation result, your setting is not the one we used to report zero-shot results, however, the performance should not differ much from the reported figures according to our evaluation (we did test on M=16 and N=16). What is the average zero-shot acc and rob of your checkpoint? Did you conduct the in-distribution evaluation to validate, if so what are the acc and rob?

**Comment by qmechlab** [comment]
Hi @TreeLLi ,
1) I haven't done any in-distribution experiments yet. I've only tried some of zeroshot's experiments so far, and here are the results for the three datasets (in parentheses are the results for the weights you just provided).
	                             Clean	             Rob
Caltech101	   75.21 (79.88)	45.72 (46.45)
OxfordFlowers   26.43 (29.07)	8.81 (9.99)
OxfordPets	   63.34 (65.25)	21.15 (21.89)
2) I would like to ask you about how the weights used for zero shot (which you just provided) are set up and trained?

**Comment by TreeLLi** [comment]
Hi,

For point 1, your results are slightly different from the numbers in the paper. May I know how you have done the evaluation? like did you use our evaluation script? if so, what is the command you run? Besides, have u followed the instruction to configurate the train/test split of the datasets?

For point 2, it used 100 shots and the context length is 1. The others are just the same.

**Comment by qmechlab** [comment]
Hi, Thank you for your reply. I admire your rigorous scientific attitude.

For 1, yes, I found that too. I conducted more experiments. I found that after loading the provided weights, the experimental results recorded in dist_shift by FGVCAircraft and OxfordFlowers are identical, other datasets may be slightly different (I posted the dist_shift content at the bottom). The path to the evaluation file I ran was APT-main/apt/evaluate. I was configured to follow the given zeroshot evaluation command (python evaluate.py path_to_checkpoint --dataset OxfordFlowers --cls-prompt prompter --attack pgd). But I didn't use the command line directly. I implemented it by changing the default item directly in pycharm (for easy debugging) as shown here. 
<img width="876" alt="Snipaste_2024-11-17_20-00-16" src="https://github.com/user-attachments/assets/4fed38cd-ba7a-4508-bc35-945e15db8d99">
The dataset as well as the json file I downloaded by following the link you provided (https://github.com/TreeLLi/APT/blob/main/DATASETS.md). I guess the program will cut it according to the configuration of the json file？

For 2, thanks for the answer, I will do this experiment next.

dist_shift file content:
Caltech101:
  prompter:
    clean: 79.87829616152007
    pgd: 46.450304275110334
DescribableTextures:
  prompter:
    clean: 24.64539005964924
    pgd: 11.524822693908186
EuroSAT:
  prompter:
    clean: 16.88888888888889
    pgd: 11.160493827160494
FGVCAircraft:
  prompter:
    clean: 7.140714071407141
    pgd: 0.7800780078007801
Food101:
  prompter:
    clean: 23.739273927392738
    pgd: 3.8382838283828384
OxfordFlowers:
  prompter:
    clean: 29.07023957624602
    pgd: 9.987819727387702
OxfordPets:
  prompter:
    clean: 65.24938672269258
    pgd: 21.88607247540528
SUN397:
  prompter:
    clean: 32.841309823677584
    pgd: 7.45088161209068
UCF101:
  prompter:
    clean: 35.474491165770196
    pgd: 7.5601374609521415

**Comment by TreeLLi** [comment]
Thanks for your appreciation. According to your description, I think you are doing right on evaluation and dataset split. Based on the attached results, the difference is trivial, which is probably due to the different random seed and the round operation.

**Comment by qmechlab** [comment]
Thank you for answering my query. By the way, I noticed that the zero shot pre-training weights are only performed for 20 epochs, so I should use the vit_b32_ep20.yaml configuration file, right?

**Comment by TreeLLi** [comment]
Yes, we used 20 epochs for 100 shots on ImageNet.

**Comment by qmechlab** [comment]
I got it. Thank you for your patience.
